DoSomething/gladiator
# User api

#### What's this PR do?
This PR adds ability for API POST requests to add users to the system and assign them to appropriate waiting rooms based on the `campaign_id` and `campaign_run_id` that they signed up for. It also restructures a bit so that both the Web CRUD and API can take advantage of the same methods via a new `Registrar` service (similar to how Northstar handles things) specific to the app. This allows DRYing up the code so while the routes where a User might be created might be different, the mechanisms under the hood are the same. 

It also updates the `UserRequest` validation so that depending on whether the request is coming from the web app itself or the API the validation rules can slightly change based on the requirements.

This PR also adds some initial API documentation in the repo, so it can easily be updated as endpoints get added/updated! Following the pattern @DFurnes established in Northstar :wink: 

#### How should this be manually tested?
Pull down, `composer dump-autoload`. Sign in and then create a new user. Does that work? Great! Next open up your favorite API HTTP Client and send a POST request to `http://gladiator.dev/api/v1/users` based on the documentation included in this PR and see what the response is!

#### Any background context you want to provide?
@sbsmith86 based on some convos with @DFurnes I changed up how we will be moving forward using the `hasAccountInSystem()` and `hasNorthstarAccount()` methods. I left the old static ones in so it won't break stuff, but planning to remove them. Instead use the new `Registrar` class to handle all that stuff! It's cleaner and more universal now, so both the Web app and the API can easily harness those methods from their respective controllers :dancer: 

#### What are the relevant tickets?
Fixes #22 

#### Next steps
- Need to create an `env` variable with a token that Phoenix will need to pass so not everyone can just sit there sending POST requests and assigning users to competitions willy-nilly.
- Fractal package has been included in project, to to apply it's use to transform the responses.


---
@angaither @sbsmith86 @deadlybutter @DFurnes
